### PR TITLE
Fetch twitter handle, profile, nft counts inside StoreCreator for creators' preview

### PR DESCRIPTION
**Pull request by alchemistgo87**
Addresses: https://github.com/holaplex/marketplace/issues/167

Updates StoreCreator object:
- Add twitter_handle
- Add profile (TwitterProfile)
- Add stats (StoreCreatorStats)

Adds dataloader:
- collection_stats_loader

Adds model & object:
- StoreCreatorStats

### CODE REVIEW

**Review by ray-kast — APPROVED** [review]

**Review by imabdulbasit — APPROVED** [review]

File: crates/graphql/src/schema/objects/store_creator.rs [hunk 1]
```diff
@@ -18,25 +19,53 @@ impl StoreCreator {
...
+            .map_err(Into::into)
+    }
+
+    pub async fn stats(&self, context: &AppContext) -> FieldResult<Option<StoreCreatorStats>> {
+        context
+            .collection_stats_loader
+            .load(self.creator_address.clone().into())
+            .await
+            .map_err(Into::into)
+    }
```
Comment (kespinola) [inline crates/graphql/src/schema/objects/store_creator.rs:52]: stats isn't quite right. We want counts for the collection previews which your query is doing but not a sales state.
```suggestion
    pub async fn nft_count(&self, context: &AppContext) -> FieldResult<Option<i32>> {
        context
            .collection_count_loader
            .load(self.creator_address.clone().into())
            .await
            .map_err(Into::into)
    }
```

File: crates/graphql/src/schema/dataloaders/stats.rs [hunk 2]
```diff
@@ -38,3 +39,19 @@ impl TryBatchFn<PublicKey<StoreConfig>, Option<MarketStats>> for Batcher {
...
+    ) -> TryBatchMap<PublicKey<StoreCreator>, Option<StoreCreatorStats>> {
+        let db = self.db()?;
+        let rows = stats::store_creator(&db, addresses)?;
+
+        Ok(rows
+            .into_iter()
+            .map(|s| (s.store_creator.clone(), s.try_into()))
+            .batch(addresses))
+    }
+}
```
Comment (kespinola) [inline crates/graphql/src/schema/dataloaders/stats.rs:57]: Lets move the data loader to store_creators.
```suggestion
#[async_trait]
impl TryBatchFn<PublicKey<StoreCreator>, Option<StoreCreatorStats>> for Batcher {
    async fn load(
        &mut self,
        addresses: &[PublicKey<StoreCreator>],
    ) -> TryBatchMap<PublicKey<StoreCreator>, Option<StoreCreatorStats>> {
        let db = self.db()?;
        let rows = stats::store_creator(&db, addresses)?;

        Ok(rows
            .into_iter()
            .map(|s| (s.store_creator.clone(), s.try_into()))
            .batch(addresses))
    }
}
```
```suggestion
#[async_trait]
impl TryBatchFn<PublicKey<StoreCreator>, Option<StoreCreatorStats>> for Batcher {
    async fn load(
        &mut self,
        addresses: &[PublicKey<StoreCreator>],
    ) -> TryBatchMap<PublicKey<StoreCreator>, Option<i64>> {
        let db = self.db()?;
        let rows = nft_counts::store_creators(&db, addresses)?;

        Ok(rows
            .into_iter()
            .map(|s| (s.store_creator.clone(), s.try_into()))
            .batch(addresses))
    }
}
```

File: crates/core/src/db/queries/stats.rs [hunk 3]
```diff
@@ -139,3 +139,30 @@ pub fn collection(
...
+const STORE_CREATOR_QUERY: &str = r"
+select
+    sc.creator_address as store_creator,
+    count(distinct mc.metadata_address)::bigint as nfts
+
+from store_creators sc
+    inner join metadata_creators mc
+        on (mc.creator_address = sc.creator_address)
+
+where sc.creator_address = any($1)
```
Comment (kespinola) [inline crates/core/src/db/queries/stats.rs:152]:
```suggestion
where sc.creator_address = any($1) AND mc.verfied
```

Only count if the metadata_creator is verified on the nft.

File: crates/core/src/db/queries/stats.rs [hunk 4]
```diff
@@ -139,3 +139,30 @@ pub fn collection(
...
+/// This function fails if the underlying SQL query returns an error
+pub fn store_creator(
+    conn: &Connection,
+    store_creators: impl ToSql<Array<Text>, Pg>,
+) -> Result<Vec<StoreCreatorStats>> {
+    diesel::sql_query(STORE_CREATOR_QUERY)
+        .bind(store_creators)
+        .load(conn)
+        .context("Failed to load marketplace stats")
+}
```
Comment (kespinola) [inline crates/core/src/db/queries/stats.rs:168]: Lets move to nft_counts queries. Its a count not a sales state like the rest of these.

File: crates/core/src/db/queries/stats.rs [hunk 5]
```diff
@@ -139,3 +139,30 @@ pub fn collection(
...
+/// # Errors
+/// This function fails if the underlying SQL query returns an error
+pub fn store_creator(
+    conn: &Connection,
+    store_creators: impl ToSql<Array<Text>, Pg>,
+) -> Result<Vec<StoreCreatorStats>> {
+    diesel::sql_query(STORE_CREATOR_QUERY)
+        .bind(store_creators)
+        .load(conn)
+        .context("Failed to load marketplace stats")
```
Comment (kespinola) [inline crates/core/src/db/queries/stats.rs:167]:
```suggestion
        .context("Failed to load store creators counts")
```

File: crates/core/src/db/queries/stats.rs [hunk 6]
```diff
@@ -139,3 +139,30 @@ pub fn collection(
...
+
+where sc.creator_address = any($1)
+group by sc.creator_address;
+ -- $1: store creator addresses::text[]";
+
+/// Count the number of nfts created by a creator
+///
+/// # Errors
+/// This function fails if the underlying SQL query returns an error
+pub fn store_creator(
```
Comment (kespinola) [inline crates/core/src/db/queries/stats.rs:160]:
```suggestion
pub fn store_creators(
```

File: crates/graphql/src/schema/dataloaders/stats.rs [hunk 7]
```diff
@@ -38,3 +39,19 @@ impl TryBatchFn<PublicKey<StoreConfig>, Option<MarketStats>> for Batcher {
...
 }
+
+#[async_trait]
+impl TryBatchFn<PublicKey<StoreCreator>, Option<StoreCreatorStats>> for Batcher {
+    async fn load(
+        &mut self,
+        addresses: &[PublicKey<StoreCreator>],
+    ) -> TryBatchMap<PublicKey<StoreCreator>, Option<StoreCreatorStats>> {
+        let db = self.db()?;
+        let rows = stats::store_creator(&db, addresses)?;
```
Comment (kespinola) [inline crates/graphql/src/schema/dataloaders/stats.rs:50]:
```suggestion
        let rows = nft_counts::store_creators(&db, addresses)?;
```

**Review by kespinola — CHANGES_REQUESTED** [review]
Updates to move around counts.

File: crates/graphql/src/schema/objects/store_creator.rs (hunk 1 above)
Comment (alchemistgo87) [reply crates/graphql/src/schema/objects/store_creator.rs:52]: @kespinola Here If we are converting FieldResult type from Option<StoreCreatorStats> to Option<i32>, then do we require the dataloader anymore? Should not we also remove the dataloader in this case?

**Review by ray-kast — APPROVED** [review]

**Review by kespinola — APPROVED** [review]

**Review by kespinola — APPROVED** [review]

**Pull request merged**